### PR TITLE
Implement Oak C++ client creation

### PR DESCRIPTION
**Pull request by ipetr0v**
Ref https://github.com/project-oak/oak/issues/4069

### CODE REVIEW

File: cc/client/client.cc [hunk 1]
```diff
@@ -28,31 +28,10 @@
...
-  absl::Status verification_status = verifier.Verify(endorsed_evidence->attestation_evidence(),
-                                                     endorsed_evidence->attestation_endorsement());
-  if (!verification_status.ok()) {
-    return verification_status;
-  }
-
-  // TODO(#4069): Pass `Transport` and enclave encryption public key to the constructor.
-  return absl::OkStatus();
-}
+namespace {}  // namespace
```
Comment (tiziano88) [inline cc/client/client.cc:31]: why is this needed?

File: cc/client/client.h [hunk 2]
```diff
@@ -34,18 +35,34 @@ class OakClient {
  public:
   // Create an instance of the Oak Client by remotely attesting an Oak Enclave
   // and creating an encrypted channel.
-  static absl::StatusOr<std::unique_ptr<OakClient>>
-  // TODO(#4069): Make `transport` a reference to a class that inherits both `EvidenceProvider` and
-  // `Transport`.
-  Create(std::unique_ptr<::oak::transport::EvidenceProvider> transport,
-         ::oak::remote_attestation::AttestationVerifier& verifier);
+  // Template class `T` should implement `::oak::transport::Transport` and
```
Comment (tiziano88) [inline cc/client/client.h:38]:
```suggestion
  // Template class `T` must implement `::oak::transport::Transport` and
```

**Review by tiziano88 — APPROVED** [review]

File: cc/client/client.cc (hunk 1 above)
Comment (ipetr0v) [reply cc/client/client.cc]: Removed

File: cc/client/client.h (hunk 2 above)
Comment (ipetr0v) [reply cc/client/client.h]: Done

File: cc/client/client.h [hunk 3]
```diff
@@ -34,18 +35,34 @@ class OakClient {
...
+  // `::oak::transport::EvidenceProvider`.
```
[the remark this answers is not in the document]
Comment (ipetr0v) [reply cc/client/client.h]: Done

File: cc/client/client.h [hunk 4]
```diff
@@ -34,18 +35,34 @@ class OakClient {
...
+  template <class T>
+  static absl::StatusOr<std::unique_ptr<OakClient>> Create(
+      std::unique_ptr<T> transport, ::oak::remote_attestation::AttestationVerifier& verifier) {
+    absl::StatusOr<::oak::session::v1::AttestationBundle> endorsed_evidence =
+        transport->GetEvidence();
+    if (!endorsed_evidence.ok()) {
```
[the remark this answers is not in the document]
Comment (ipetr0v) [reply cc/client/client.h:43]: These macros are not open-sourced =(

File: cc/client/client.h [hunk 5]
```diff
@@ -34,18 +34,33 @@ class OakClient {
  public:
   // Create an instance of the Oak Client by remotely attesting an Oak Enclave
   // and creating an encrypted channel.
-  static absl::StatusOr<std::unique_ptr<OakClient>>
-  // TODO(#4069): Make `transport` a reference to a class that inherits both `EvidenceProvider` and
-  // `Transport`.
-  Create(std::unique_ptr<::oak::transport::EvidenceProvider> transport,
-         ::oak::remote_attestation::AttestationVerifier& verifier);
+  template <class T>
```
[the remark this answers is not in the document]
Comment (ipetr0v) [reply cc/client/client.h]: Removed

File: cc/client/client.h [hunk 6]
```diff
@@ -34,7 +34,6 @@ class OakClient {
  public:
   // Create an instance of the Oak Client by remotely attesting an Oak Enclave
   // and creating an encrypted channel.
-  template <class T>
   static absl::StatusOr<std::unique_ptr<OakClient>> Create(
       std::unique_ptr<::oak::transport::TransportWrapper> transport,
       ::oak::remote_attestation::AttestationVerifier& verifier) {
```
[the remark this answers is not in the document]
Comment (ipetr0v) [reply cc/client/client.h]: Done

**Review by pmcgrath17 — APPROVED** [review]
Thanks!
